### PR TITLE
Bag-of-words featurizer for the states

### PR DESCRIPTION
**Pull request by evgeniiaraz**
**Proposed changes**:
- created a new `SingleStateFeaturizer` `BOWSingleStateFeaturizer` which creates a BOW representation of current state; 
(it builds on top of `CountVectorsFeaturizer` from `nlu`)
- `_postprocess_trackers_as_states` function in `TrackerFeaturizer`which gets rid of repetitive intents across turns;

**Status (please check what you already did)**:
- [x] added some tests for the functionality (on the way)
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by Ghostvv** [comment]
to enforce the change, could you please remove the other ones

**Comment by evgeniiaraz** [comment]
What do you mean by "removing the other ones"? the other changes?

**Comment by Ghostvv** [comment]
remove other featurizers

File: rasa/core/featurizers.py [hunk 1]
```diff
@@ -133,6 +137,76 @@ def create_encoded_all_actions(self, domain: Domain) -> np.ndarray:
...
+        """
+
+        training_data = domain.input_states
+        training_data = [
+            Message(key.replace(delimiter, " ") + " CLS") for key in training_data
+        ]
+        training_data = TrainingData(training_examples=training_data)
+        self.train(training_data)
+
+    def encode(self, state: Dict[Text, float], delimiter="_", type_output="numpyarray"):
```
Comment (Ghostvv) [inline rasa/core/featurizers.py]: let's make delimiter a class variable

File: rasa/core/featurizers.py [hunk 2]
```diff
@@ -133,6 +137,76 @@ def create_encoded_all_actions(self, domain: Domain) -> np.ndarray:
         return np.eye(domain.num_actions)
 
 
+class BOWSingleStateFeaturizer(CountVectorsFeaturizer, SingleStateFeaturizer):
+    def prepare_training_data_and_train(self, trackers_as_states, delimiter="_"):
```
Comment (Ghostvv) [inline rasa/core/featurizers.py]: could you please add a TODO that for real text data there will be no delimiter, and we'll need to think about utilizing whole nlu pipeline. But for now current implementation should be enough

File: rasa/core/featurizers.py [hunk 3]
```diff
@@ -133,6 +137,76 @@ def create_encoded_all_actions(self, domain: Domain) -> np.ndarray:
...
+             - trackers_as_states: real data as a dictionary
+             - delimiter: symbol to be used to divide into words
+        """
+        training_data = []
+        for tracker in trackers_as_states:
+            for state in tracker:
+                if state:
+                    state_keys = list(state.keys())
+                    state_keys = [
+                        Message(key.replace(delimiter, " ") + " CLS")
```
Comment (Ghostvv) [inline rasa/core/featurizers.py]: let's use `CLS` constant variable

File: rasa/core/featurizers.py [hunk 4]
```diff
@@ -133,6 +137,76 @@ def create_encoded_all_actions(self, domain: Domain) -> np.ndarray:
...
+    def prepare_from_domain(self, domain: Domain, delimiter="_") -> None:
+        """
+        Train the fecturizers based on the inputs gotten from the domain;
+        Args:
+            - domain: domain file
+        """
+
+        training_data = domain.input_states
+        training_data = [
+            Message(key.replace(delimiter, " ") + " CLS") for key in training_data
```
Comment (Ghostvv) [inline rasa/core/featurizers.py]: let's use `CLS` constant variable

File: rasa/core/featurizers.py (hunk 1 above)
Comment (Ghostvv) [inline rasa/core/featurizers.py]: let's use `dense` and `sparse` for `type_output`

File: rasa/core/featurizers.py [hunk 5]
```diff
@@ -133,6 +137,76 @@ def create_encoded_all_actions(self, domain: Domain) -> np.ndarray:
...
+        Args:
+            - state: dictionary describing current state
+            - type output: type to return the features as (numpyarray or sklearn coo_matrix)
+        Returns:
+            - nparray(vocab_size,) or coo_matrix(1, vocab_size) 
+        """
+
+        if state is None:
+            return (
+                np.ones(len(self.vectorizers["text"].vocabulary_), dtype=np.int32) * -1
```
Comment (Ghostvv) [inline rasa/core/featurizers.py]: should be sparse for sparse output

File: rasa/core/featurizers.py [hunk 6]
```diff
@@ -133,6 +137,76 @@ def create_encoded_all_actions(self, domain: Domain) -> np.ndarray:
...
+            )
+
+        state_keys = [key.replace(delimiter, " ") for key in list(state.keys())]
+        attribute = "text"
+        message = Message(" ".join(state_keys))
+        message_tokens = self._get_processed_message_tokens_by_attribute(
+            message, attribute
+        )
+
+        message_tokens = " ".join(message_tokens)
```
Comment (Ghostvv) [inline rasa/core/featurizers.py]: there is no need to join message_tokens, the last output (for `CLS` token) will contain bag of word features

File: rasa/core/featurizers.py [hunk 7]
```diff
@@ -133,6 +137,76 @@ def create_encoded_all_actions(self, domain: Domain) -> np.ndarray:
...
+        # features shape (1, seq, dim)
```
Comment (Ghostvv) [inline rasa/core/featurizers.py]: it should be `(batch, seq, dim)`

File: rasa/core/featurizers.py [hunk 8]
```diff
@@ -376,6 +452,27 @@ def _featurize_labels(
 
         return y
 
+    def _postprocess_trackers_as_states(
```
Comment (Ghostvv) [inline rasa/core/featurizers.py]: is it simpler to check `if PREV_PREFIX + ACTION_LISTEN_NAME in state:` otherwise remove intent?

**Pull request closed without merging**